org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -0,0 +1,781 @@
+        }
+
+        if (vrange == null) {
+            // Again; this shouldn't happen as the vrange should always
+            // be created by VersionRangeGet(...).
+            throw new RuntimeException("JSSEngine.queryEnabledProtocols() - null protocol range");
+        }
+
+        min_protocol = vrange.getMinVersion();
+        max_protocol = vrange.getMaxVersion();
+    }
+
+    /**
+     * Gets the list of enabled SSL protocol versions on this particular
+     * JSSEngine instance, as a list of JCA-standardized strings.
+     */
+    public String[] getEnabledProtocols() {
+        logger.debug("JSSEngine: getEnabledProtocols()");
+
+        if (min_protocol == null || max_protocol == null) {
+            queryEnabledProtocols();
+            assert(min_protocol != null && max_protocol != null);
+        }
+
+        // Use JSSParameters to do the heavy lifting of converting the
+        // SSLVersionRange to a list of JDK-conforming Strings.
+        JSSParameters parser = new JSSParameters();
+        parser.setProtocols(min_protocol, max_protocol);
+        return parser.getProtocols();
+    }
+
+    /**
+     * Gets the list of SSL protocols supported, as a list of JCA-standardized
+     * strings.
+     */
+    public String[] getSupportedProtocols() {
+        logger.debug("JSSEngine: getSupportedProtocols()");
+        ArrayList<String> result = new ArrayList<String>();
+
+        for (SSLVersion v : Policy.TLS_VERSION_RANGE.getAllInRange()) {
+            logger.debug("JSSEngine: getSupportedProtocol - Supported: " + v);
+            result.add(v.jdkAlias());
+        }
+
+        return result.toArray(new String[result.size()]);
+    }
+
+    /**
+     * Set public and protected key material; useful when doing client auth or
+     * if this wasn't provided to the constructor.
+     *
+     * Can also be used to remove key material; however note that both
+     * arguments must match: either both certificate and key are null or
+     * both are not-null.
+     */
+    public void setKeyMaterials(PK11Cert our_cert, PK11PrivKey our_key) throws IllegalArgumentException {
+        logger.debug("JSSEngine: setKeyMaterials()");
+
+        if ((our_cert == null && our_key != null) || (our_cert != null && our_key == null)) {
+            throw new IllegalArgumentException("JSSEngine.setKeyMaterials(): Either both cert and key must be null or both must be not-null");
+        }
+
+        cert = our_cert;
+        key = our_key;
+    }
+
+    /**
+     * Set the internal KeyManager, when present, replacing all previous
+     * KeyManagers.
+     *
+     * It is suggested that at least one key manager be a JSSKeyManager
+     * instance if a key and certificate must be provided for this end of
+     * the connection.
+     */
+    public void setKeyManager(X509KeyManager km) {
+        if (km == null) {
+            logger.debug("JSSEngine: setKeyManager(null)");
+            return;
```

Comment:
Should we set `key_managers` to `null` here?

Reply:
I don't know. I think the behavior of this will change once we have support for more TrustManagers/KeyManagers here. I don't know what all we want, but IMO, we should have something else. SNI for one.

Reply:
Generally there's two common paths:

1. You select your `SSLContext` and end up (by happenstance or choice) with a `Mozilla-JSS` `JSSContextSpi`-backed `SSLContext`. You call `.init(...)` and passed the Key/Trust managers you're expected. This is the expected procedure. `JSSContextSpi` handles passing `X509KeyManager` to the underlying `SSLEngine` implementation. You don't interact with this `SSLEngine` at all really. You only pass a `setSSLParameters(...)` which doesn't actually hold a `KeyManager`/`TrustManager` either.


But if you know you really want _this_ `SSLEngine` and plan on configuring it all yourself (or even for `JSSContext` use), I added these method for the following path:

2. Instantiate a new `JSSEngine` instance. Then you're going to need to provide it a `KeyManager`/`TrustManager`. You're going to need to configure it as you see fit (ciphers, ...). And then you can go ahead and use it.

I'm not sure removing an existing `KeyManager` is going to do you any good. It means you probably can't find keys. It means you'd have to directly pass the set of credentials you want to use to the `JSSEngine` (which'd be preferred anyways, if you pass them, regardless of what value the `KeyManager` takes).

And you'd probably always need a `TrustManager`, otherwise you'd only use NSS's built in cert authentication handler, which is very basic.

Reply:
I dunno, both are questions I hope to address later. There's still only one `JSSKeyManager`, and unless you set the key directly (which if you do, the key manager is ignored so it doesn't matter if it is `null` or not)... there's no other way of getting keys. So its a matter of "when" you fail (with a `null` `JSSKeyManager`), not "if". 

If we get more than one `JSSKeyManager` then yeah, I think we could introduce an `add()` and handle overwriting them. But until then... ~shrugs~? I don't see this as a big deal.